lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -1078,20 +1089,44 @@ ConstExprFunctionState::computeSingleStoreAddressValue(SILValue addr) {
   SingleValueInstruction *addrInst = dyn_cast<AllocStackInst>(addr);
   if (!addrInst)
     addrInst = dyn_cast<PointerToAddressInst>(addr);
-  if (!addrInst)
+  if (!addrInst) {
+    // Example SIL snippet involving index_addr that we can handle:
+    // %78 = index_addr %73 : $*Int32, %77 : $Builtin.Word // user: %86
+    // function_ref SignedInteger<>.init<A>(_:)
+    // %85 = function_ref @... // user: %86
+    // Here the func call writes to %78.
+    // %86 = apply %85<Int32, Int>(%78, %83, %80) : $@convention(method)
     addrInst = dyn_cast<IndexAddrInst>(addr);
+  }
+  if (!addrInst) {
+    if (auto *teai = dyn_cast<TupleElementAddrInst>(addr)) {
+      // In this case, we simply const-evaluate this inst, which involves first
+      // const-evaluating the tuple operand of this inst (to an aggregate const
+      // value). e.g. %131 below.
+      // %132 = tuple_element_addr %131 : $*(Int32, Int32, Int32, Int32), 0
+      auto tupleEltAddr = computeConstantValue(addr);
+      return tupleEltAddr;
+    }
+  }
   if (!addrInst)
```

Comment:
This approach predated the current approach to handling addresses in the constexpr stuff.  Instead of pattern matching the SIL instructions directly, how about just calling:

auto addrCst = getConstantValue(addr);

and then handling the various forms of address constants?  That will allow handling recursive forms of indexing.

One other random point: generally you shouldn't call computeConstantValue directly, because this skips caching.  Is there a reason you're not calling getConstantValue?

Reply:
I studied the design and code base more, and came to the belief that we'll need two different code paths for handling the (fairly small) set of inst types that return address values, such as `pointer_to_address`, `index_addr` and `tuple_element_addr`. 

e.g. given an inst like
```
%132 = tuple_element_addr %131 : $*(Int32, Int32, Int32, Int32), 0
```

When we are reading from `%132`, we should const-evaluating %131 into a tuple const first, and use that result to const-eval `%132`. But if we are writing into `%132` (as handled by the code here), we will be const-evaluating a writer of `%132`, as a way to const-evaluate `%131`. The flow of information differs between these two cases. Fortunately, the chunk of code in `getConstValue()` that handles these addr-typed insts is fairly small, so there is not much added complexity.

For more context, PTAL the comment block above `getSingleWriterAddressValue()`, and how it contrasts with `getConstAddrAndLoadResult()` (I renamed `computeLoadResult()` to the latter, which is based on `getConstantValue()`.